DeviceAdapters/PyDevice/PyCamera.cpp

```diff
@@ -0,0 +1,252 @@
+* exact dimensions requested - but should try do as close as possible.
+* If the hardware does not have this capability the software should simulate the ROI by
+* appropriately cropping each frame.
+*/
+int CPyCamera::SetROI(unsigned x, unsigned y, unsigned xSize, unsigned ySize)
+{
+    if (xSize == 0 && ySize == 0) // special case: reset ROI
+        return ClearROI();
+
+    // apply ROI
+    PyLock lock; // make sure all four elements of the ROI are set without any other thread having access in between
+    SetLongProperty(g_Keyword_Left, x);
+    SetLongProperty(g_Keyword_Top, y);
+    SetLongProperty(g_Keyword_Width, xSize);
+    SetLongProperty(g_Keyword_Height, ySize);
+    return DEVICE_OK;
+}
+
+/**
+* Returns the actual dimensions of the current ROI.
+* If multiple ROIs are set, then the returned ROI should encompass all of them.
+* Required by the MM::Camera API.
+*/
+int CPyCamera::GetROI(unsigned& x, unsigned& y, unsigned& xSize, unsigned& ySize)
+{
+    PyLock lock; // make sure all four elements of the ROI are read without any other thread having access
+    x = GetLongProperty(g_Keyword_Left);
+    y = GetLongProperty(g_Keyword_Top);
+    xSize = GetLongProperty(g_Keyword_Width);
+    ySize = GetLongProperty(g_Keyword_Height);
+    return DEVICE_OK;
+}
+
+/**
+* Resets the Region of Interest to full frame.
+* Required by the MM::Camera API.
+*/
+int CPyCamera::ClearROI()
+{
+    PyLock lock; // make sure all four elements of the ROI are set without any other thread having access in between
+    double width, height, top, left;
+    GetPropertyLowerLimit(g_Keyword_Top, top);
+    GetPropertyLowerLimit(g_Keyword_Left, left);
+    GetPropertyUpperLimit(g_Keyword_Width, width);
+    GetPropertyUpperLimit(g_Keyword_Height, height);
+    SetLongProperty(g_Keyword_Top, static_cast<long>(top));
+    SetLongProperty(g_Keyword_Left, static_cast<long>(left));
+    SetLongProperty(g_Keyword_Width, static_cast<long>(width));
+    SetLongProperty(g_Keyword_Height, static_cast<long>(height));
+    return DEVICE_OK;
+}
+
+
+/**
+* Returns the current exposure setting in milliseconds.
+* Required by the MM::Camera API.
+*/
+double CPyCamera::GetExposure() const
+{
+    return GetFloatProperty(g_Keyword_Exposure);
+}
+
+/**
+* Sets exposure in milliseconds.
+* Required by the MM::Camera API.
+*/
+void CPyCamera::SetExposure(double exp_ms)
+{
+    if (SetFloatProperty(g_Keyword_Exposure, exp_ms) == DEVICE_OK)
+        GetCoreCallback()->OnExposureChanged(this, exp_ms);
+}
+
+/**
+* Returns the current binning factor. Currently only a binning of 1 (no binning) is supported
+* Required by the MM::Camera API.
+*/
+int CPyCamera::GetBinning() const
+{
+    return GetLongProperty(g_Keyword_Binning);
+}
+
+/**
+* Sets binning factor.
+* Required by the MM::Camera API.
+*/
+int CPyCamera::SetBinning(int binF)
+{
+    return SetLongProperty(g_Keyword_Binning, binF);
+}
+
+int CPyCamera::IsExposureSequenceable(bool& isSequenceable) const
+{
```

Comment:
This should default to False. Exposure sequencing is an advanced feature that most cameras do not support